packages/graph-editor/src/components/controls/array.tsx

```diff
@@ -0,0 +1,178 @@
+
+const inputItemTypes = [STRING, NUMBER, COLOR];
+
+export const ArrayField = observer(({ port }: IField) => {
+    const [value, setValue] = React.useState(port.value);
+    const [selectItemsType, setSelectItemsType] = React.useState(port.type.items.$id);
+
+    const [autofocusIndex, setAutofocusIndex] = React.useState<number | null>(null);
+    const useDelayed = useSelector(delayedUpdateSelector);
+
+    React.useEffect(() => {
+        setValue(port.value);
+    }, [port.value]);
+
+    React.useEffect(() => {
+        if (autofocusIndex !== null) {
+            setAutofocusIndex(null);
+        }
+    }, [autofocusIndex]);
+
+    const portName = port.constructor.name;
+    const itemsType = port.type.items.$id;
+
+    const onColorChange =
+        (e: React.ChangeEvent<HTMLInputElement>, index: number) => {
+            let newColor: string;
+            // Weird  problem with the color picker if the user decides to use the text input
+            if (typeof e === 'string') {
+                newColor = e
+            } else {
+                newColor = e.target.value;
+            }
+
+            onChange(newColor, index);
+        };
+
+    const onChange = (newValue: any, index: number) => {
+        const newValueArray = value.map((val: any, i: number) => {
+            if (i === index) {
+                return itemsType === NUMBER ? +newValue : newValue;
+            }
+            return val;
+        });
+        setValue(newValueArray);
+
+        if (useDelayed) {
+            return;
+        }
+
+        (port as Input).setValue(newValueArray);
+    };
+
+    const addItem = () => {
+        const newValueArray = [...value];
+
+        switch (itemsType) {
+            case STRING:
+                newValueArray.push('');
+                break;
+            case NUMBER:
+                newValueArray.push(undefined);
+                break;
+            case COLOR:
+                newValueArray.push('#000000');
+                break;
+            default:
+                newValueArray.push('');
+                break;
+        }
+
+
+        setValue(newValueArray);
+        (port as Input).setValue(newValueArray);
+
+        setAutofocusIndex(newValueArray.length - 1);
+    };
+
+    const removeItem = (index: number) => {
+        const newValueArray = value.filter((val: any, i: number) => i !== index);
+
+        setValue(newValueArray);
+        (port as Input).setValue(newValueArray);
+
+        setAutofocusIndex(null);
+    }
+
+    const setType = () => {
+        const schema = AllSchemas.find((x) => x.$id === selectItemsType);
+        if (!schema) {
+            return;
+        }
+
+        const input = port as Input;
+        input.setValue(value, {
+            type: {
+                type: "array",
+                items: schema,
+            },
+        });
+    }
+
+    if (!port.isConnected && portName === 'Input' && itemsType === ANY && !(port as Input).variadic) {
```

Comment:
Why are we checking that the port name is Input ?

Reply:
Because we do not want to show anything of this for output ports. This is only for inputs